src/core/model/table/dynamic_position_list_index.cpp

```diff
@@ -0,0 +1,176 @@
+            cluster.insert(cluster.end(), new_cluster.begin(), new_cluster.end());
+            new_clusters.erase(new_cluster_value);
+        }
+
+        if (cluster.empty()) {
+            it = clusters_.erase(it);
+        } else {
+            sort(cluster.begin(), cluster.end());
+            ++it;
+        }
+    }
+
+    for (auto& new_cluster : new_clusters) {
+        clusters_.emplace(std::move(new_cluster));
+    }
+}
+
+std::unordered_map<int, unsigned> DynamicPositionListIndex::CreateFrequencies(
+        Cluster const& cluster, std::vector<int> const& probing_table) {
+    std::unordered_map<int, unsigned> frequencies;
+
+    for (int index : cluster) {
+        frequencies[probing_table[index]]++;
+    }
+
+    return frequencies;
+}
+
+std::shared_ptr<std::vector<int> const> DynamicPositionListIndex::CalculateAndGetProbingTable()
+        const {
+    if (probing_table_cache_ != nullptr) return probing_table_cache_;
+    cluster_values_cache_.clear();
+    std::vector<int> probing_table = std::vector<int>(next_record_id_);
+    int next_cluster_id = 1;
+    cluster_values_cache_.push_back({});
+    for (auto& [value, cluster] : clusters_) {
+        int value_id = next_cluster_id++;
+        for (int position : cluster) {
+            probing_table[position] = value_id;
+        }
+        cluster_values_cache_.emplace_back(value);
+    }
+
+    return std::make_shared<std::vector<int>>(probing_table);
+}
+
+std::unique_ptr<DynamicPositionListIndex> DynamicPositionListIndex::Intersect(
+        DynamicPositionListIndex const* that) const {
+    assert(this->next_record_id_ == that->next_record_id_);
+
+    if (this->valid_records_number_ > that->valid_records_number_) {
+        return that->Probe(this);
+    } else {
+        return this->Probe(that);
+    }
+}
+
+std::unique_ptr<DynamicPositionListIndex> DynamicPositionListIndex::Probe(
+        DynamicPositionListIndex const* that) const {
+    std::shared_ptr<std::vector<int> const> probing_table = that->CalculateAndGetProbingTable();
+    std::vector<ClusterValue> const& cluster_values = that->GetClusterValuesCache();
+    assert(this->next_record_id_ == probing_table->size());
+    ClusterCollection new_clusters;
+    unsigned int new_size = 0;
+
+    std::unordered_map<int, std::pair<ClusterValue, Cluster>> partial_clusters;
+
+    for (auto& [value, cluster] : clusters_) {
+        for (int position : cluster) {
+            int probing_table_value_id = (*probing_table)[position];
+            if (partial_clusters.contains(probing_table_value_id)) {
+                partial_clusters[probing_table_value_id].second.emplace_back(position);
+            } else {
+                ClusterValue key = std::move(value);
+                key.insert(key.end(), cluster_values[probing_table_value_id].begin(),
+                           cluster_values[probing_table_value_id].end());
+                partial_clusters[probing_table_value_id] = {std::move(key), {position}};
```

Comment:
Repeated lookup for `partial_clusters` and `cluster_values`. Using `try_emplace` and storing references will help avoid this.

Reply:
I don't think, that try_emplace is a good idea as in this case we will construct key every time

Reply:
`partial_clusters.try_emplace(probing_table_value_id, std::piecewise_construct, std::forward_as_tuple(std::move(v)), std::forward_as_tuple(std::initializer_list{position}));`